src/single_random_draw.rs

```diff
@@ -0,0 +1,166 @@
+//! This library provides efficient algorithms to compose a set of unspent transaction outputs
+//! (UTXOs).
+
+#[cfg(any(test, feature = "rand"))]
+use crate::errors::LibError;
+
+use bitcoin::Amount;
+use bitcoin::blockdata::transaction;
+use bitcoin::FeeRate;
+use bitcoin::TxOut;
+use bitcoin::Weight;
+use crate::CHANGE_LOWER;
+use rand::seq::SliceRandom;
+
+/// Randomly select coins for the given target by shuffling the utxo pool and
+/// taking UTXOs until the given target is reached.
+#[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    outputs: &mut Vec<TxOut>,
+    fee_rate: FeeRate,
+    rng: &mut R)
+-> Result<Option<Vec<TxOut>>, LibError> {
+    outputs.shuffle(rng);
+
+    let mut selected_effective_value: Amount = Amount::ZERO;
+    let mut i = 0;
+    let mut result = Vec::new();
+
+    // The required size of a change output must be at least
+    // larger than the target by CHANGE_LOWER.  Therefore,
+    // CHANGE_LOWER + target becomes the new lower bound for the
+    // selection.
+    while selected_effective_value < target + CHANGE_LOWER && i < outputs.len() {
+        let output: &TxOut = &outputs[i];
+        let value: Amount = output.value;
+
+        let outputs = outputs.iter().map(|txout| txout.script_pubkey.len());
+
+        let weight: Weight = transaction::predict_weight(
+            None::<Vec<transaction::InputWeightPrediction>>,
+            Some(outputs),
```

Comment:
Shouldn't it be checking the fees for inputs rather than outputs? (selecting an additional utxo will add an input to the transaction, but maybe I'm misunderstanding how this function is supposed to work).

Reply:
hmm that's a good point, and I think you are correct.  The owner of the outputs will transition them to be inputs for the transaction, and so, when included in a transaction, they are actually now inputs instead of outputs, which changes the weight calculation.

Reply:
Thanks, I believe this has been resolved now.